rfd/0131-adminitrative-actions-mfa.md

```diff
@@ -0,0 +1,488 @@
+for automated use cases).
+
+### Proto Specification
+
+Each Admin action API request will need to include the `MFAAuthenticateResponse`
+field in its proto specification.
+
+Clients will automatically check for the `MFAAuthenticateResponse` in order to
+determine whether it must prompt for MFA. This will be done seamlessly within
+a `grpc.UnaryInterceptor`.
+
+Likewise, the Server will automatically check for an `MFAAuthenticateResponse` in
+using a `grpc.UnaryInterceptor`. If found, this will be passed down the chain
+through the `AuthContext` (like user) where it can be verified by `ServerWithRoles`
+if required.
+
+#### HTTP Endpoints
+
+Some requests which we would like to make admin actions are still used through
+the HTTP API rather than the gRPC API, such as `POST UpsertUser`. Part of this
+work will require converting these requests to gRPC, which will also move us
+closer to [fully deprecating the HTTP API](https://github.com/gravitational/teleport/issues/6394).
+
+#### Non-unique gRPC Requests
+
+Many any of our existing gRPC API requests do not use unique request or response
+messages, despite it being a best practice. As a result, it is not trivial to add
+the `MFAAuthenticateResponse` field to some requests.
+
+For example:
+
+```proto
+rpc UpdateUser(types.UserV2) returns (google.protobuf.Empty);
+```
+
+For these cases, we will need to create replacement requests:
+
+```proto
+rpc UpdateUser(types.UserV2) returns (google.protobuf.Empty);
+rpc UpdateUserV2(UpdateUserV2Request) returns (UpdateUserV2Response);
+```
+
+For many of these cases above, it will make more sense to add or move the rpc
+into a new service. This will allow us to:
+
+- Maintain preferred naming conventions, avoiding `V2` suffixes
+- Reduce reliance on gogoproto which has got to go
+- Reduce size of legacy proto packages which are incompatible with buf
+
+This will be handled on a case by case basis, but user service looks like a
+likely candidate:
+
+```proto
+service UserService {
+  rpc GetUser(GetUserRequest) returns (types.UserV2);
+  rpc GetUsers(GetUsersRequest) returns (stream types.UserV2);
+  // We can replace CreateUser and UpdateUser with UpsertUser, as is the latest convention of our API.
+  rpc UpsertUser(UpsertUserRequest) returns (UpsertUserResponse);
+  rpc DeleteUser(DeleteUserRequest) returns (google.protobuf.Empty);
+  ...
+}
+```
+
+### Test Plan
+
+The implementation of this feature will include automated unit tests to ensure
+that MFA is required for admin actions across applicable server configurations.
+
+### Security
+
+The bulk of this RFD is focused on security and will result in a net positive
+to our security outlook.
+
+Here are a few key points to review:
+
+- [Server Configuration](#server-configuration):
+  - In order to maintain backwards compatibility and non-MFA
+  compatible use cases (automated plugins), we give server admins the option to
+  turn off MFA for admin actions for the cluster or specific roles. This also
```

Comment:
In order to support backwards compatibility, do we need to allow the role level disable to disable in all cases?

One specific case we could benefit from is requiring MFA always for role changes.  Otherwise we need to continue to consider possible MFA bypass from any auth which currently has a role that disables MFA.

If we can limit the scope of what can be disabled it will make it safer in the long run.

Reply:
All of the resources which we particularly want to put behind MFA are managed by the [Terraform provider](https://goteleport.com/docs/reference/terraform-provider/), so I don't think we can limit the MFA bypass for specified endpoints across the board.

However, we could make the bypass resource-scoped. This way MFA bypass would only apply to resources explicitly set for the role. Something like:

```
kind: role
metadata:
  name: mfa-bypass-for-roles
spec:
  allow:
    db_labels:
      '*': '*'
    app_labels:
      '*': '*'
    rules:
      - resources:
        - role
        verbs: ['list','create','read','update','delete','mfa-bypass']
```

Reply:
> As I understand it, this configuration is primarily for our built in roles, so hopefully this configuration complexity is not something our customers would generally need to burden themselves with.

Not quite, these are two separate problems. Built in roles are used for our internals: `RoleProxy`, `RoleAuth`, `RoleNode`, etc. There is also `RoleAdmin` which is used exclusively for running `tctl` commands directly on Auth server. These roles are used directly for RBAC as opposed to Teleport users with a set of roles, if that makes sense. All of these internal roles will be exempt from MFA.
 
On the other hand, If you take a look at the [Terraform Provider](https://goteleport.com/docs/management/dynamic-resources/terraform-provider/#create-the-terraform-user), it used a normal user+role setup for RBAC. So in this example we would update the role setup to include the `mfa-bypass` verb in the terraform role:

```yaml
kind: role
metadata:
  name: terraform
spec:
  allow:
    db_labels:
      '*': '*'
    app_labels:
      '*': '*'
    rules:
      - resources:
        - app
        - cluster_auth_preference
        - cluster_networking_config
        - db
        - device
        - github
        - login_rule
        - oidc
        - okta_import_rule
        - role
        - saml
        - session_recording_config
        - token
        - trusted_cluster
        - user
        verbs: ['list','create','read','update','delete','mfa-bypass']
version: v6
---
kind: user
metadata:
  name: terraform
spec:
  roles: ['terraform']
version: v2
```

> so hopefully this configuration complexity is not something our customers would generally need to burden themselves with.

This should be the case, as customers will only need to configure `mfa-bypass` for specific use cases according to our guides, aka the plugin guides.

Reply:
As @codingllama [pointed out](https://github.com/gravitational/teleport/pull/27588#discussion_r1270681363), I don't think adding a `bypass-mfa` verb is a great idea after all, due to the complexity and configuration bloat.

Additionally, in the example I shared above, there isn't really much difference between having the MFA bypass setting as a `role_option`. The MFA bypass applies to any of the admin actions the Terraform Plugin can do either way.